### DISCUSSION TITLE
Get jupyter-matlab-proxy to work with Quarto?

### DISCUSSION THREAD

**Discussion by allefeld** [quarto-dev/quarto-cli#6195] [Q&A]
### Description

A few weeks ago, Mathworks released an official Jupyter kernel for Matlab. It would be great if this kernel could be used with Quarto, but so far I can't get it to work, see https://github.com/mathworks/jupyter-matlab-proxy/issues/48.

I'm pretty sure that this is something that needs to be solved on the side of `jupyter-matlab-proxy`, but just in case I wanted to let you know about this.

**Comment by mcanouil** [comment]
FYI, Quarto documentation on kernel selection: <https://quarto.org/docs/computations/python.html#kernel-selection>.

If you manage to use it in a Jupyter notebook, there is no (apparent) reason that it won't work in Quarto.

**Comment by cderv** [comment]
For the upstream issue, I see this comment 
> It seems like the tool you are using is starting MATLAB Kernel directly instead of using a Jupyter server because of which you are seeing the error.

Really not sure about it, but can you check or try with running as a daemon. 
https://quarto.org/docs/computations/python.html#kernel-daemon

This should be the default, but maybe not in your case (if on Windows for example I think).

Probably makes no difference - just want to check.  I do think we start kernel directly with Quarto and it seems somehow their Matlab Kernel does not support that... I don't enough yet of Jupyter internal to see the difference.

**Reply by allefeld** [reply]
@cderv, I just tried with `execute:\n  daemon: true`, and no, it makes no difference. I'm on Debian GNU/Linux 11.7, so it shouldn't.

Thanks for your comment on the other issue! I think you're right that the fail point is NBClient. I added further comments to explain how to reproduce the issue, and also advertised a bit for supporting Quarto. Let's hope they enable direct usage of the kernel in a future version.

On this side, is there any way to make Quarto/NBClient use a Jupyter server instead of the kernel directly? I couldn't find anything about that in the NBClient documentation though.

**Reply by cderv** [reply]
> On this side, is there any way to make Quarto/NBClient use a Jupyter server instead of the kernel directly? I couldn't find anything about that in the NBClient documentation though.

Sorry I don't know. Probably something to discuss with Nbclient folks, and Matlab kernel author. Then we could see to adapt here if needed.

**Reply by allefeld** [reply]
I just saw, there is also [Jupyter Client](https://jupyter-client.readthedocs.io/en/stable/), I guess Quarto could use that instead of NBClient.

**Reply by cderv** [reply]
[`nbclient` is using Jupyter Client](https://github.com/jupyter/nbclient/blob/30a013a7519723b43ffef2cc9a47d8254cf838da/pyproject.toml#L38). So we are already using it, but through `nbclient`. I don't think this will change as we need the `nbclient` layer (for some reason I can't explain as I don't know this part enough - sorry)

Main difference is highlighted at https://github.com/jupyter/nbclient#relationship-to-jupyterclient 
> While, nbclient allows notebooks to be run in different execution contexts.

This is key I believe.
